### ISSUE TITLE
Interrupt service routine with callback 

### ISSUE THREAD

**Issue by kgbkgb** [wiringX/wiringX#83]
I am planning to use the OrangePi PC computer, Armbian 5.14, to implement a system with ADC, DACs and a counter to replace an outdated MSDOS system I have built
I need to use an interrupt input to make a counter, exactly as in the "isr.c" example in WiringPi.. This works fine with WiringPi on a Rasberrypi 2. I tried this using WiringOP on the OrangePi PC (H3 cpu) but it is flaky.
WiringX works better on the OrangePi but the interrupt support seems to only allow a "wait for interrupt" routine.
WiringPi also has a "link to user supplied interrupt service routine upon interrupt"
int wiringPiISR (int pin, int mode, void (*function)(void))
I need to use this type of interrupt routine to avoid the polling or blocking read in the "wait for interrupt".
Is there some problem preventing this implementation in WiringX?
I am not a system programmer, but should i be able to implement this function by "pasting" across the int wiringPiISR function from WiringPi and attempting to fix variable names etc?
(Linux User-land interrupts are OK as the frequency is <300Hz and not response-time critical)

**Comment by CurlyMoo** [comment]
Can you post the code of your current implementation (interrupt part) with wiringPi?

**Comment by kgbkgb** [comment]
This code which I mentioned is straight from WiringPi, file wiringPi.c.
Although this is present in the wiringOP conversion, it does not work - it compiles but the counter example program isr.c gets no counts. A wait-for-interrupt example on the same header pin can detect interrupts. The wiringOP version is rather confusing with Banana Pi modifications also.
EDIT: wait for interrupt does not  work using WiringOP, only using WiringX.  Using the same pin for IO and the blink.c example code does work.

The wiringOP package is here   [https://github.com/zhaolei/WiringOP](url)
I also include the example file isr.c from WiringPi, which uses this interrupt routine to provide a counter of interrupt events.

====  extract from wiringPi.c=======

```
int wiringPiISR (int pin, int mode, void (*function)(void))
{
  pthread_t threadId ;
  const char *modeS ;
  char fName   [64] ;
  char  pinS [8] ;
  pid_t pid ;
  int   count, i ;
  char  c ;
  int   bcmGpioPin ;

  if ((pin < 0) || (pin > 63))
    return wiringPiFailure (WPI_FATAL, "wiringPiISR: pin must be 0-63 (%d)\n", pin) ;

  /**/ if (wiringPiMode == WPI_MODE_UNINITIALISED)
    return wiringPiFailure (WPI_FATAL, "wiringPiISR: wiringPi has not been initialised. Unable to continue.\n") ;
  else if (wiringPiMode == WPI_MODE_PINS)
    bcmGpioPin = pinToGpio [pin] ;
  else if (wiringPiMode == WPI_MODE_PHYS)
    bcmGpioPin = physToGpio [pin] ;
  else
    bcmGpioPin = pin ;
// Now export the pin and set the right edge
//  We're going to use the gpio program to do this, so it assumes
//  a full installation of wiringPi. It's a bit 'clunky', but it
//  is a way that will work when we're running in "Sys" mode, as
//  a non-root user. (without sudo)

  if (mode != INT_EDGE_SETUP)
  {
    /**/ if (mode == INT_EDGE_FALLING)
      modeS = "falling" ;
    else if (mode == INT_EDGE_RISING)
      modeS = "rising" ;
    else
      modeS = "both" ;

    sprintf (pinS, "%d", bcmGpioPin) ;

    if ((pid = fork ()) < 0)    // Fail
      return wiringPiFailure (WPI_FATAL, "wiringPiISR: fork failed: %s\n", strerror (errno)) ;

    if (pid == 0)   // Child, exec
    {
      /**/ if (access ("/usr/local/bin/gpio", X_OK) == 0)
      {
    execl ("/usr/local/bin/gpio", "gpio", "edge", pinS, modeS, (char *)NULL) ;
    return wiringPiFailure (WPI_FATAL, "wiringPiISR: execl failed: %s\n", strerror (errno)) ;
      }
      else if (access ("/usr/bin/gpio", X_OK) == 0)
      {
    execl ("/usr/bin/gpio", "gpio", "edge", pinS, modeS, (char *)NULL) ;
    return wiringPiFailure (WPI_FATAL, "wiringPiISR: execl failed: %s\n", strerror (errno)) ;
      }
      else
    return wiringPiFailure (WPI_FATAL, "wiringPiISR: Can't find gpio program\n") ;
    }
    else        // Parent, wait
      wait (NULL) ;
  }

// Now pre-open the /sys/class node - but it may already be open if
//  we are in Sys mode...

  if (sysFds [bcmGpioPin] == -1)
  {
    sprintf (fName, "/sys/class/gpio/gpio%d/value", bcmGpioPin) ;
    if ((sysFds [bcmGpioPin] = open (fName, O_RDWR)) < 0)
      return wiringPiFailure (WPI_FATAL, "wiringPiISR: unable to open %s: %s\n", fName, strerror (errno)) ;
  }

// Clear any initial pending interrupt

  ioctl (sysFds [bcmGpioPin], FIONREAD, &count) ;
  for (i = 0 ; i < count ; ++i)
    read (sysFds [bcmGpioPin], &c, 1) ;

  isrFunctions [pin] = function ;

  pthread_mutex_lock (&pinMutex) ;
    pinPass = pin ;
    pthread_create (&threadId, NULL, interruptHandler, NULL) ;
    while (pinPass != -1)
      delay (1) ;
  pthread_mutex_unlock (&pinMutex) ;

  return 0 ;
}
```

==== end of extract from wirinPi.c ==============

===== complete isr.c example from wiringPi package =====

```
/*
 * isr.c:
 *  Wait for Interrupt test program - ISR method
 *
 *  How to test:
 *    Use the SoC's pull-up and pull down resistors that are avalable
 *  on input pins. So compile & run this program (via sudo), then
 *  in another terminal:
 *      gpio mode 0 up
 *      gpio mode 0 down
 *  at which point it should trigger an interrupt. Toggle the pin
 *  up/down to generate more interrupts to test.
 *
 * Copyright (c) 2013 Gordon Henderson.
 ***********************************************************************
 * This file is part of wiringPi:
 *  https://projects.drogon.net/raspberry-pi/wiringpi/
 *
 *    wiringPi is free software: you can redistribute it and/or modify
 *    it under the terms of the GNU Lesser General Public License as published by
 *    the Free Software Foundation, either version 3 of the License, or
 *    (at your option) any later version.
 *
 *    wiringPi is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU Lesser General Public License for more details.
 *
 *    You should have received a copy of the GNU Lesser General Public License
 *    along with wiringPi.  If not, see <http://www.gnu.org/licenses/>.
 ***********************************************************************
 */

#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <stdlib.h>
#include <wiringPi.h>


// globalCounter:
//  Global variable to count interrupts
//  Should be declared volatile to make sure the compiler doesn't cache it.

static volatile int globalCounter [8] ;


/*
 * myInterrupt:
 *********************************************************************************
 */

void myInterrupt0 (void) { ++globalCounter [0] ; }
void myInterrupt1 (void) { ++globalCounter [1] ; }
void myInterrupt2 (void) { ++globalCounter [2] ; }
void myInterrupt3 (void) { ++globalCounter [3] ; }
void myInterrupt4 (void) { ++globalCounter [4] ; }
void myInterrupt5 (void) { ++globalCounter [5] ; }
void myInterrupt6 (void) { ++globalCounter [6] ; }
void myInterrupt7 (void) { ++globalCounter [7] ; }


/*
 *********************************************************************************
 * main
 *********************************************************************************
 */

int main (void)
{
  int gotOne, pin ;
  int myCounter [8] ;

  for (pin = 0 ; pin < 8 ; ++pin) 
    globalCounter [pin] = myCounter [pin] = 0 ;

  wiringPiSetup () ;

  wiringPiISR (0, INT_EDGE_FALLING, &myInterrupt0) ;
  wiringPiISR (1, INT_EDGE_FALLING, &myInterrupt1) ;
  wiringPiISR (2, INT_EDGE_FALLING, &myInterrupt2) ;
  wiringPiISR (3, INT_EDGE_FALLING, &myInterrupt3) ;
  wiringPiISR (4, INT_EDGE_FALLING, &myInterrupt4) ;
  wiringPiISR (5, INT_EDGE_FALLING, &myInterrupt5) ;
  wiringPiISR (6, INT_EDGE_FALLING, &myInterrupt6) ;
  wiringPiISR (7, INT_EDGE_FALLING, &myInterrupt7) ;

  for (;;)
  {
    gotOne = 0 ;
    printf ("Waiting ... ") ; fflush (stdout) ;

    for (;;)
    {
      for (pin = 0 ; pin < 8 ; ++pin)
      {
    if (globalCounter [pin] != myCounter [pin])
    {
      printf (" Int on pin %d: Counter: %5d\n", pin, globalCounter [pin]) ;
      myCounter [pin] = globalCounter [pin] ;
      ++gotOne ;
    }
      }
      if (gotOne != 0)
    break ;
    }
  }

  return 0 ;
}
```

**Comment by CurlyMoo** [comment]
I of course know what wiringPi implemented, but i want to know how you used the function in your own code.

**Comment by kgbkgb** [comment]
Sorry CurlyMoo. I have not started coding this new project yet. I am still at the hardware feasibility point and without an interrupt service feature the project is not feasible on an OrangePi - i would have to use a genuine raspberryPi.  I did use an interrupt vector routine in my 1990 equipment, but that was TurboC on MSDOS and I am trying to escape from that evil environment!  ;-)
All i need is the isr.c  example code capability: a global counter of interrupt events. Events come from a detector and amplifier, with pre-conditioning  in CMOS hardware to prevent overlapping events etc. That counter will be read at regular intervals using the system timer somehow.

**Comment by CurlyMoo** [comment]
Just start using wiringX and when you get stuck because i implemented the interrupt callback differently, come back again and i'll help you.

**Comment by kgbkgb** [comment]
I dont understand how to start coding without having an interrupt function which has a callback to a user supplied interrupt service routine. Have I missed something here? In WiringX i can only find a "waitforInterrupt" function. This blocks awaiting an interrupt doesn't it? But I need the code to continue graphing the data and controlling the DACs and reading the ADC while interrupts are being counted by the ISR routine without blocking. I think i need to use the    wiringPiISR (BUTTON_PIN, INT_EDGE_FALLING, &myInterrupt)    type of function as in WiringPi, where the &myinterrupt function increments the interrupt counter and returns control to the main code while awaiting the next interrupts.
Are you suggesting there is a way to use waitforinterrupt () without needing to run a separate thread?

**Comment by CurlyMoo** [comment]
> Are you suggesting there is a way to use waitforinterrupt () without needing to run a separate thread?

No there isn't but have you actually looked under the hood of the wiringPiISR:

```
    pthread_create (&threadId, NULL, interruptHandler, NULL) ;
```

That's why it's a stupid implementation because they force you to use their threaded implementation instead of your own.

**Comment by kgbkgb** [comment]
I am having major issues trying to get interrupts to work on the orangepi-PC.
In wiringX file orangepi.c, there is the following tabulation:

```
static int edge [64] =
{
  -1, -1, -1, -1, 4, -1, -1, 7,   //support the INT
  8, 9, 10, 11, -1,-1, 14, 15,
  -1, 17, -1, -1, -1, -1, 22, 23,
  24, 25, -1, 27, -1, -1, -1, -1,      ....etc.
```

I tested all the GPIO pins on this list, but most of them fail with:
orangepi: error while calling isr
these GPIO# pins are 11, 14, 0, 3, 4, 5, 6, 2 - they all give this fail error.
Only 6 pins are accepted by the wiringx-interrupt 5 x example/test program:
these are GPIO#  7, 10, 13, 12, 15, 16
But none detected the interrupt trigger I used (which I tested on a raspberry pi 2, and which worked properly)
AT the end of these tests i found the following GPIO files in /sys/class/gpio:  10, 198, 199, 200, 201, 68, 7, 8, 9
I used wiringx-blink to check that pins were changing voltage, to confirm my connections.

To cross check, I used the wiringOP package. (yes i know about the library conflict between these packages and removed the conflicting libraries for each test). I think there is a major issue with pin number mappings and only GPIO# 0, 7, 14 work for interrupt input, but these fail to trigger on an input signal. I suspect the mode selections for edge triggering are badly broken, probably as this package was converted first to a bananapi with A20 cpu, then to orangepi with H3 cpu.

I used a RaspberryPi model 2 with wiringPi and it worked correctly and confirmed that my interrupt input circuit is OK. 

In my early wiringX tests I thought i did get edge triggering working with the wiringx-wfi example program, but i cannot repeat this test. And the pin number mapping confusion means only one or 2 pins are available as I am using I2C and also SPI devices.
At present I will probably change to use Raspberry hardware as it and wiringPi work fine. I would still like to fix the int issues on wiringx on orangepi-PC, but right now i am too frustrated.
